python/lsst/pipe/base/graph/_versionDeserializers.py

```diff
@@ -0,0 +1,524 @@
+            # taskDef back into the un-persisted node
+            object.__setattr__(qNode, 'taskDef', loadedTaskDef[nodeTask])
+            quanta[qNode.taskDef].add(qNode.quantum)
+
+            # record the node for later processing
+            quantumToNodeId[qNode.quantum] = qNode.nodeId
+
+        # construct an empty new QuantumGraph object, and run the associated
+        # creation method with the un-persisted data
+        qGraph = object.__new__(QuantumGraph)
+        qGraph._buildGraphs(quanta, _quantumToNodeId=quantumToNodeId, _buildId=self.returnValue._buildId,
+                            metadata=self.returnValue.metadata)
+        return qGraph
+
+    def description(self) -> str:
+        return Version1Description
+
+
+Version2Description = """
+The save file starts with the first few bytes corresponding to the magic bytes
+in the QuantumGraph: `qgraph4\xf6\xe8\xa9`.
+
+The next few bytes are a big endian unsigned long long.
+
+The unsigned long long corresponds to the number of bytes of a python mapping
+of header information. This mapping is encoded into json and then lzma
+compressed, meaning the operations must be performed in the opposite order to
+deserialize.
+
+The json encoded header mapping contains 4 fields: TaskDefs, GraphBuildId,
+Nodes, and Metadata.
+
+The `TaskDefs` key corresponds to a value which is a mapping of Task label to
+task data. The task data is a mapping of key to value, where the only key is
+`bytes` and it corresponds to a tuple of a byte range of the start, stop
+bytes (indexed after all the header bytes)
+
+The `GraphBuildId` corresponds with a string that is the unique id assigned to
+this graph when it was created.
+
+The `Nodes` key is like the `TaskDefs` key except it corresponds to
+QuantumNodes instead of TaskDefs. Another important difference is that JSON
+formatting does not allow using numbers as keys, and this mapping is keyed by
+the node number. Thus it is stored in JSON as two equal length lists, the first
+being the keys, and the second the values associated with those keys.
+
+The `Metadata` key is a mapping of strings to associated values. This metadata
+may be anything that is important to be transported alongside the graph.
+
+As stated above, each map contains byte ranges of the corresponding
+datastructure. Theses bytes are also lzma compressed pickles, and should
+be deserialized in a similar manor.
+"""
+
+
+@dataclass
+class DeserializerV2(DeserializerBase):
+    FMT_STRING: ClassVar[str] = '>Q'
+
+    def __post_init__(self):
+        self.mapSize, = struct.unpack(self.FMT_STRING, self.sizeBytes)
+
+    @property
+    def headerSize(self) -> int:
+        return self.preambleSize + self.structSize + self.mapSize
+
+    def readHeaderInfo(self, rawHeader: bytes) -> SimpleNamespace:
+        uncompressedHeaderMap = self.unpackHeader(rawHeader)
+        assert uncompressedHeaderMap is not None  # This deserializer can never have None option
```

Comment:
not actually an io error, its a type checker thing since the signature on the base class takes into account all possible subclass signatures, but this class can never return None. But I changed it to an instance check instead of an assert.